File: frontend/src/api/songs.ts
import { apiClient, ApiError } from './client'
import type { AlignedWord } from '@/types/lyrics'

// Music Style enum with 8 preset styles
export enum MusicStyle {
  POP = 'pop',
  RAP = 'rap',
  FOLK = 'folk',
  ELECTRONIC = 'electronic',
  ROCK = 'rock',
  JAZZ = 'jazz',
  CHILDREN = 'children',
  CLASSICAL = 'classical',
}

// TypeScript interfaces
export interface GenerateSongRequest {
  lyrics: string
  style: MusicStyle
  content_hash?: string
}

export interface GenerateSongResponse {
  task_id: string
  estimated_time: number // seconds
}

/**
 * Represents a single song variation from Suno API
 * Requirements: 1.1, 7.1
 */
export interface SongVariation {
  audio_url: string
  audio_id: string
  variation_index: number // 0 or 1
}

/**
 * Song generation status update
 * Requirements: 7.2, 7.4
 */
export interface SongStatusUpdate {
  task_id: string
  status: 'queued' | 'processing' | 'completed' | 'failed'
  progress: number // 0-100
  song_url?: string // Deprecated, use variations
  variations: SongVariation[] // Array of song variations (1-2 items)
  error?: string
}

// Retry configuration
const MAX_RETRIES = 3
const RETRY_DELAY = 1000 // 1 second

/**
 * Helper function to retry API calls with exponential backoff
 */
async function retryWithBackoff<T>(
  fn: () => Promise<T>,
  retries: number = MAX_RETRIES
): Promise<T> {
  try {
    return await fn()
  } catch (error) {
    // Only retry if error is retryable and we have retries left
    if (error instanceof ApiError && error.retryable && retries > 0) {
      const delay = RETRY_DELAY * Math.pow(2, MAX_RETRIES - retries)
      await new Promise(resolve => setTimeout(resolve, delay))
      return retryWithBackoff(fn, retries - 1)
    }
    throw error
  }
}

// API functions
export const generateSong = async (
  request: GenerateSongRequest
): Promise<GenerateSongResponse> => {
  return retryWithBackoff(() =>
    apiClient.post<GenerateSongResponse>('/api/songs/generate', request)
  )
}

export const getSongStatus = async (
  taskId: string
): Promise<SongStatusUpdate> => {
  return retryWithBackoff(() =>
    apiClient.get<SongStatusUpdate>(`/api/songs/${taskId}`)
  )
}

/**
 * TEST FUNCTION: Generate song with simulated timeout
 * This calls a special backend endpoint that simulates a 90+ second timeout
 * for testing timeout handling in the frontend.
 */
export const generateSongWithTimeout = async (
  request: GenerateSongRequest
): Promise<GenerateSongResponse> => {
  // No retry for timeout test - we want to see the timeout behavior
  return apiClient.post<GenerateSongResponse>('/api/songs/generate-timeout-test', request)
}

// Song Details interfaces for playback page
/**
 * Complete song details for playback page
 * Requirements: 7.2, 7.4
 */
export interface SongDetails {
  song_id: string
  song_url: string // Deprecated, use variations
  variations: SongVariation[] // Array of song variations (1-2 items)
  primary_variation_index: number // Index of user's selected primary variation (0 or 1)
  lyrics: string
  style: MusicStyle
  created_at: string  // ISO datetime
  expires_at: string  // ISO datetime
  is_owner: boolean
  // Timestamped lyrics fields
  aligned_words?: AlignedWord[] | null
  waveform_data?: number[] | null
  has_timestamps?: boolean
}

export interface ShareLinkResponse {
  share_url: string
  share_token: string
  expires_at: string  // ISO datetime
}

/**
 * Get complete song details for playback page
 * @param songId - The ID of the song to retrieve
 * @returns SongDetails with song_url, lyrics, style, metadata, and expiration info
 * @throws ApiError with 404 if song not found, 410 if expired, 403 if unauthorized
 */
export const getSongDetails = async (songId: string): Promise<SongDetails> => {
  return retryWithBackoff(() =>
    apiClient.get<SongDetails>(`/api/songs/${songId}/details`)
  )
}

/**
 * Get song details via share token (no auth required)
 * @param shareToken - The share token from the shared link
 * @returns SongDetails for the shared song
 * @throws ApiError with 404 if share token not found, 410 if share link expired
 */
export const getSharedSong = async (shareToken: string): Promise<SongDetails> => {
  return retryWithBackoff(() =>
    apiClient.get<SongDetails>(`/api/songs/shared/${shareToken}`)
  )
}

/**
 * Generate a shareable link for a song
 * @param songId - The ID of the song to share
 * @returns ShareLinkResponse with share_url and expires_at
 * @throws ApiError with 403 if user doesn't own the song, 404 if song not found
 */
export const createShareLink = async (songId: string): Promise<ShareLinkResponse> => {
  return retryWithBackoff(() =>
    apiClient.post<ShareLinkResponse>(`/api/songs/${songId}/share`)
  )
}

/**
 * Update the user's primary song variation selection
 * Requirements: 4.1, 7.5
 * @param taskId - Song task ID
 * @param variationIndex - Index of selected variation (0 or 1)
 * @returns Success response with updated primary_variation_index
 * @throws ApiError with 403 if user doesn't own the song, 404 if song not found, 400 if invalid index
 */
export const updatePrimaryVariation = async (
  taskId: string,
  variationIndex: number
): Promise<{ success: boolean; primary_variation_index: number }> => {
  return retryWithBackoff(() =>
    apiClient.patch<{ success: boolean; primary_variation_index: number }>(
      `/api/songs/${taskId}/primary-variation`,
      { variation_index: variationIndex }
    )
  )
}

/**
 * Fetch timestamped lyrics for a specific variation
 * Requirements: 6.1
 * @param taskId - Song task ID
 * @param variationIndex - Which variation (0 or 1)
 * @returns Timestamped lyrics data with aligned_words and waveform_data
 * @throws ApiError with 403 if user doesn't own the song, 404 if song not found, 400 if invalid index
 */
export const fetchVariationTimestampedLyrics = async (
  taskId: string,
  variationIndex: number
): Promise<{ aligned_words: AlignedWord[]; waveform_data: number[] }> => {
  return retryWithBackoff(() =>
    apiClient.post<{ aligned_words: AlignedWord[]; waveform_data: number[] }>(
      `/api/songs/${taskId}/timestamped-lyrics/${variationIndex}`
    )
  )
}

/**
 * Song history summary for display in history list
 * Requirements: 6.2
 */
export interface SongHistorySummary {
  song_id: string
  style: MusicStyle
  created_at: string  // ISO datetime
  expires_at: string  // ISO datetime
  lyrics_preview: string  // First 100 characters
  has_variations: boolean
  primary_variation_index: number
}

/**
 * Get user's song history (non-expired songs)
 * Requirements: 6.1
 * @returns List of SongHistorySummary ordered by created_at DESC
 * @throws ApiError with 401 if unauthenticated
 */
export const getSongHistory = async (): Promise<SongHistorySummary[]> => {
  return retryWithBackoff(() =>
    apiClient.get<SongHistorySummary[]>('/api/songs/history')
  )
}

File: backend/app/models/songs.py
"""Pydantic models for song generation."""

from datetime import datetime
from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field, field_validator


class MusicStyle(str, Enum):
    """Available music styles for song generation."""
    
    POP = "pop"
    RAP = "rap"
    FOLK = "folk"
    ELECTRONIC = "electronic"
    ROCK = "rock"
    JAZZ = "jazz"
    CHILDREN = "children"
    CLASSICAL = "classical"


class GenerateSongRequest(BaseModel):
    """Request model for song generation."""
    
    lyrics: str = Field(
        ...,
        description="Song lyrics to generate music for",
        min_length=50,
        max_length=3000
    )
    style: MusicStyle = Field(
        ...,
        description="Music style for the generated song"
    )
    content_hash: Optional[str] = Field(
        default=None,
        description="SHA-256 hash of the original content for caching"
    )
    
    @field_validator('lyrics')
    @classmethod
    def validate_lyrics(cls, v: str) -> str:
        """Validate lyrics are not empty or whitespace only."""
        if not v or not v.strip():
            raise ValueError('Lyrics cannot be empty')
        
        stripped = v.strip()
        if len(stripped) < 50:
            raise ValueError(
                f'Lyrics must be at least 50 characters (current: {len(stripped)} characters)'
            )
        if len(stripped) > 3000:
            raise ValueError(
                f'Lyrics exceed 3000 character limit (current: {len(stripped)} characters)'
            )
        
        return v


class GenerateSongResponse(BaseModel):
    """Response model for song generation initiation."""
    
    task_id: str = Field(
        ...,
        description="Unique identifier for the song generation task"
    )
    estimated_time: int = Field(
        ...,
        description="Estimated time to complete generation in seconds",
        ge=0
    )


class GenerationStatus(str, Enum):
    """Status of a song generation task."""
    
    IDLE = "idle"
    QUEUED = "queued"
    PROCESSING = "processing"
    COMPLETED = "completed"
    FAILED = "failed"


class SongVariation(BaseModel):
    """Represents a single song variation from Suno API.
    
    Requirements: 1.1, 7.1
    """
    
    audio_url: str = Field(
        ...,
        description="URL of the song audio file"
    )
    audio_id: str = Field(
        ...,
        description="Unique identifier for fetching timestamped lyrics"
    )
    variation_index: int = Field(
        ...,
        description="Index of this variation (0 or 1)",
        ge=0,
        le=1
    )


class SongStatusUpdate(BaseModel):
    """Model for song generation status updates (WebSocket messages).
    
    Requirements: 7.2, 7.4
    """
    
    task_id: str = Field(
        ...,
        description="Unique identifier for the song generation task"
    )
    status: GenerationStatus = Field(
        ...,
        description="Current status of the generation task"
    )
    progress: int = Field(
        default=0,
        description="Progress percentage (0-100)",
        ge=0,
        le=100
    )
    song_url: Optional[str] = Field(
        default=None,
        description="URL of the generated song (when completed) - deprecated, use variations"
    )
    variations: list[SongVariation] = Field(
        default_factory=list,
        description="Array of song variations (1-2 items)"
    )
    error: Optional[str] = Field(
        default=None,
        description="Error message (when failed)"
    )


class AlignedWordDict(BaseModel):
    """Represents a word with timing information for lyrics synchronization.
    
    Requirements: 2.2, 2.3
    """
    
    word: str = Field(
        ...,
        description="The word or phrase text"
    )
    startS: float = Field(
        ...,
        description="Start time in seconds",
        ge=0.0
    )
    endS: float = Field(
        ...,
        description="End time in seconds",
        ge=0.0
    )
    success: bool = Field(
        default=True,
        description="Whether alignment was successful"
    )
    palign: float = Field(
        default=0.0,
        description="Alignment confidence score",
        ge=0.0,
        le=1.0
    )
    
    @field_validator('endS')
    @classmethod
    def validate_end_after_start(cls, v: float, info) -> float:
        """Validate that endS is not less than startS."""
        start_s = info.data.get('startS')
        if start_s is not None and v < start_s:
            raise ValueError('endS cannot be less than startS')
        return v


class SongDetails(BaseModel):
    """Complete song details for playback page.
    
    Requirements: 8.1, 8.4, 2.2, 2.3, 7.2, 7.4
    """
    
    song_id: str = Field(
        ...,
        description="Unique identifier for the song"
    )
    song_url: str = Field(
        ...,
        description="URL of the generated song audio file - deprecated, use variations"
    )
    variations: list[SongVariation] = Field(
        default_factory=list,
        description="Array of song variations (1-2 items)"
    )
    primary_variation_index: int = Field(
        default=0,
        description="Index of user's selected primary variation",
        ge=0,
        le=1
    )
    lyrics: str = Field(
        ...,
        description="Song lyrics text"
    )
    style: MusicStyle = Field(
        ...,
        description="Music style of the song"
    )
    created_at: datetime = Field(
        ...,
        description="Timestamp when the song was created"
    )
    expires_at: datetime = Field(
        ...,
        description="Timestamp when the song will expire"
    )
    is_owner: bool = Field(
        ...,
        description="True if the requesting user owns the song"
    )
    # Timestamped lyrics fields (Requirements: 2.2, 2.3)
    aligned_words: Optional[list[dict]] = Field(
        default=None,
        description="Array of aligned words with timing information"
    )
    waveform_data: Optional[list[float]] = Field(
        default=None,
        description="Waveform data for visualization"
    )
    has_timestamps: bool = Field(
        default=False,
        description="Whether timestamped lyrics are available"
    )
    
    model_config = {
        "json_encoders": {
            datetime: lambda v: v.isoformat()
        }
    }


class UpdatePrimaryVariationRequest(BaseModel):
    """Request to update primary variation selection.
    
    Requirements: 4.1, 7.5
    """
    
    variation_index: int = Field(
        ...,
        description="Index of the variation to set as primary (0 or 1)",
        ge=0,
        le=1
    )


class SongHistorySummary(BaseModel):
    """Summary of a song for history list.
    
    Requirements: 6.2
    """
    
    song_id: str = Field(
        ...,
        description="Unique identifier for the song"
    )
    style: MusicStyle = Field(
        ...,
        description="Music style of the song"
    )
    created_at: datetime = Field(
        ...,
        description="Timestamp when the song was created"
    )
    expires_at: datetime = Field(
        ...,
        description="Timestamp when the song will expire"
    )
    lyrics_preview: str = Field(
        ...,
        description="First 100 characters of lyrics",
        max_length=100
    )
    has_variations: bool = Field(
        ...,
        description="Whether the song has multiple variations"
    )
    primary_variation_index: int = Field(
        ...,
        description="Index of the user's selected primary variation",
        ge=0,
        le=1
    )
    
    model_config = {
        "json_encoders": {
            datetime: lambda v: v.isoformat()
        }
    }


class ShareLinkResponse(BaseModel):
    """Response for share link creation.
    
    Requirements: 5.1, 5.2
    """
    
    share_url: str = Field(
        ...,
        description="Full URL for sharing the song"
    )
    share_token: str = Field(
        ...,
        description="Unique token for the share link"
    )
    expires_at: datetime = Field(
        ...,
        description="Timestamp when the share link will expire"
    )
    
    model_config = {
        "json_encoders": {
            datetime: lambda v: v.isoformat()
        }
    }
